### ISSUE TITLE
Binary vectors(Akaze) perform better results than real-valued vectors(Sift) using l2 distance?

### ISSUE THREAD

**Issue by igor-87** [facebookresearch/faiss#351]
Hi, I do some testing by comparing sift vectors (real-valued) vs akaze (binary).
I take about 20k images and calculate their vectors with sift/akaze descriptors.
using IndexIVFPQR as my index.
From my understanding, IndexIVFPQR uses l2 distance to get the closest results.
From my testing, it seems that akaze performed much better results than sift although it's a binary vector that should use hamming distance to get closest results.
how it's possible?

**Comment by mdouze** [comment]
Hi
What do you call "better"? Is your quality criterion image-based or with respect to exhaustive distance computations?
If you convert a bit vector to a float vector of +/- 1 or 0/1 then Hamming distance is equivalent to L2 distance.

**Comment by igor-87** [comment]
sorry I forgot to say that when I mean "better" is based on a total count of the closest vectors ids from the search results and ordering by the highest result. 
base on the tests I made in this method and computations binary vectors performed better results.
is there any function in the library that can do this conversion?
or can you please guide me with which library I can do it?  

thanks

**Comment by mdouze** [comment]
There is an example on how to use binary vectors in Faiss: 
https://github.com/facebookresearch/faiss/blob/master/tests/test_index.py#L496

**Comment by igor-87** [comment]
Hi, I am a little bit confused, I have been trying to implement the test_encode_to_binary and I have some issues.
I use akaze descriptor by my understanding its binary descriptor that the output its 56 d vector. for example
[174.0,107.0,185.0,178.0,107.0,224.0,246.0,165.0,186.0,100.0,147.0,67.0,190.0,78.0,247.0.......]
From the example of test_encode_to_binary, i see that the xb data generated as 8 d from the beginning before any ProductQuantizer.
and after the ProductQuantizer the binary_to_float function make a reshape for each vector on d*8 .
I am a little bit confused with the process maybe you can please give me a small explanation about the process and how should I do it on 56 d vectors.

**Comment by mdouze** [comment]
If you have a d-bit descriptor, it is encoded as a d/8 byte (uint8) vector. 
However, to input it to Faiss, you need a d-dim float vector.
The binary_to_float function converts between the two representations.
You seem to be using a float vector to represent the uint8's.

**Comment by igor-87** [comment]
Thank you for your response, very appreciate your help.
So is there a difference between float vector with uint8 values or float vector that contains float values In terms of the accuracy?.

Sorry for the confusing questions, I am trying to figure out what is the right technique / way for my project.
Basically, I am trying to build a small search engine for images inner product that will show As close as possible images with the same product (about 150-200k images).
My current steps are:
1)Getting images vectors using openCv
  I tried sift/akaze and its seems opencv extract for both Descriptors float vector with uint8 values.
2)Indexing all the vectors using IndexIVFPQ.
3)Searching and ordering the results by the highest id count.

By the current results, akaze performed better results than sift but I noticed that there is a lot of false positive results.

What is the right way to approach the problem I try to solve, to get the max of the similar images?
